syft/formats/common/cyclonedxhelpers/licenses.go

```diff
@@ -1,53 +1,208 @@
-	if c.Licenses != nil {
-		for _, l := range *c.Licenses {
-			if l.License != nil {
-				var lic string
-				switch {
-				case l.License.ID != "":
-					lic = l.License.ID
-				case l.License.Name != "":
-					lic = l.License.Name
-				default:
+func decodeLicenses(c *cyclonedx.Component) []pkg.License {
+	licenses := make([]pkg.License, 0)
+	if c == nil || c.Licenses == nil {
+		return licenses
+	}
+
+	for _, l := range *c.Licenses {
+		if l.License == nil {
+			continue
+		}
+		// these fields are mutually exclusive in the spec
+		switch {
+		case l.License.ID != "":
+			licenses = append(licenses, pkg.LicenseFromURLs(l.License.ID, l.License.URL))
+		case l.License.Name != "":
+			licenses = append(licenses, pkg.LicenseFromURLs(l.License.Name, l.License.URL))
+		case l.Expression != "":
+			licenses = append(licenses, pkg.LicenseFromURLs(l.Expression, l.License.URL))
+		default:
+		}
+	}
+
+	return licenses
+}
+
+// nolint:funlen
+func separateLicenses(p pkg.Package) (spdx, other cyclonedx.Licenses, expressions []string) {
+	spdxc := cyclonedx.Licenses{}
+	otherc := cyclonedx.Licenses{}
+	ex := make([]string, 0)
+	/*
+			pkg.License can be a couple of things:
+			- Complex SPDX expression
+			- Some other Valid license ID
+			- Some non standard non spdx license
+
+			To determine if an expression is a singular ID we first run it against the SPDX license list.
+
+		The weird case we run into is if there is a package with a license that is not a valid SPDX expression
+			and a license that is a valid complex expression. In this case we will surface the valid complex expression
+			as a license choice and the invalid expression as a license string.
+
+	*/
+	// dedupe spdxlicenseID
+	seen := make(map[string]bool)
+	for _, l := range p.Licenses.ToSlice() {
+		// singular expression case
+		if value, exists := spdxlicense.ID(l.SPDXExpression); exists {
+			// we do 1 license -> many URL in our internal model
+			// this fans out different URL to single cyclone licenses
+			if !l.URL.Empty() {
+				for _, url := range l.URL.ToSlice() {
+					if url != "" {
+						spdxc = append(spdxc, cyclonedx.LicenseChoice{
+							License: &cyclonedx.License{
+								ID:  value,
+								URL: url,
+							},
+						})
+						continue
+					}
+					spdxc = append(spdxc, cyclonedx.LicenseChoice{
+						License: &cyclonedx.License{
+							ID: value,
+						},
+					})
+					continue
+				}
+			}
+			if _, exists := seen[value]; exists {
+				continue
+			}
+			spdxc = append(spdxc, cyclonedx.LicenseChoice{
+				License: &cyclonedx.License{
+					ID: value,
+				},
+			})
+			seen[value] = true
+			continue
+		}
+		if l.SPDXExpression != "" {
+			// COMPLEX EXPRESSION CASE: do we instead break the spdx expression out
+			// into individual licenses OR combine singular licenses into a single expression?
+			ex = append(ex, l.SPDXExpression)
+			continue
+		}
+
+		// license string that are not valid spdx expressions or ids
+		if !l.URL.Empty() {
```

Comment:
should the continue's here be updating `seen`? or does that only apply to IDs and not expressions?